nup/sliceUpdate_test.go

```diff
@@ -0,0 +1,464 @@
+			}
+			if isSet != testCase.expectedIsSet {
+				t.Errorf("Expected isSet: %v. Actual: %v", testCase.expectedIsSet, isSet)
+			}
+		})
+	}
+}
+
+func TestSliceUpdate_ValueOrNil(t *testing.T) {
+	testCases := []struct {
+		name     string
+		update   SliceUpdate[int]
+		expected []int
+	}{
+		{
+			name:     "Noop",
+			update:   SliceNoop[int](),
+			expected: nil,
+		},
+		{
+			name:     "Remove",
+			update:   SliceRemove[int](),
+			expected: nil,
+		},
+		{
+			name:     "Set",
+			update:   SliceSet(testSlice1),
+			expected: testSlice1,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			actual := testCase.update.ValueOrNil()
+			if !reflect.DeepEqual(actual, testCase.expected) {
+				t.Errorf("Expected value: %v. Actual: %v", testCase.expected, actual)
+			}
+		})
+	}
+}
+
+func TestSliceUpdate_Apply(t *testing.T) {
+	testCases := []struct {
+		name     string
+		u        SliceUpdate[int]
+		expected []int
+	}{
+		{
+			name:     "Noop",
+			u:        SliceNoop[int](),
+			expected: testSlice1,
+		},
+		{
+			name:     "Remove",
+			u:        SliceRemove[int](),
+			expected: nil,
+		},
+		{
+			name:     "Set",
+			u:        SliceSet(testSlice2),
+			expected: testSlice2,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			if actual := testCase.u.Apply(testSlice1); !reflect.DeepEqual(actual, testCase.expected) {
+				t.Errorf("Expected: %v. Actual: %v", testCase.expected, actual)
+			}
+		})
+	}
+}
+
+func TestSliceUpdate_Diff(t *testing.T) {
+	testCases := []struct {
+		name     string
+		u        SliceUpdate[int]
+		value    []int
+		expected SliceUpdate[int]
+	}{
+		{
+			name:     "Noop",
+			u:        SliceNoop[int](),
+			value:    testSlice1,
+			expected: SliceNoop[int](),
+		},
+		{
+			name:     "Remove/NonZeroValue",
+			u:        SliceRemove[int](),
+			value:    testSlice1,
+			expected: SliceRemove[int](),
+		},
+		{
+			name:     "Remove/ZeroValue",
+			u:        SliceRemove[int](),
+			value:    nil,
+			expected: SliceNoop[int](),
+		},
+		{
+			name:     "Set/Equal",
+			u:        SliceSet(testSlice1),
+			value:    testSlice1,
+			expected: SliceNoop[int](),
+		},
+		{
+			name:     "Set/NotEqual",
+			u:        SliceSet(testSlice2),
+			value:    testSlice1,
+			expected: SliceSet(testSlice2),
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			if actual := testCase.u.Diff(testCase.value); !reflect.DeepEqual(actual, testCase.expected) {
+				t.Errorf("Expected: %v. Actual: %v", testCase.expected, actual)
+			}
+		})
+	}
+}
+
+func TestSliceUpdate_IsSetTo(t *testing.T) {
+	testCases := []struct {
+		name     string
+		u        SliceUpdate[int]
+		expected bool
+	}{
+		{
+			name:     "Noop",
+			u:        SliceNoop[int](),
+			expected: false,
+		},
+		{
+			name:     "Remove",
+			u:        SliceSet[int](nil),
```

Comment:
Very good point! I fixed those cases and added some extra tests and a comment concerning this ambiguity in 636cd1b245edb0e81fc2d6fe2a2eb4355c51cbfb. In short, I think we actually don't need to worry about it.

Removing a slice field is semantically equivalent to setting it to `nil`. This is analogous to how removing an `int` field means setting it to `0`. The only weird thing (as you alluded in the roundtrip tests) is that there's no way to unmarshal a slice update as `SliceSet[T](nil)` - because it will be unmarshalled as `SliceRemove[T]()` instead. But I can't think of any current uses in our codebase where the distinction would matter. Thoughts?

Reply:
I think this is okay -- I like that it's documented in the tests now. I'm still a little perturbed that there is some weirdness like `SliceSet[T](nil).String()` is `"[]"` but marshals (IMO correctly) to `null`. Ultimately what I'm concerned about is a gotcha cropping up from someone using `SliceSet[T](nil)`, for instances in logging when a variable of type `SliceUpdate` might be used as a stringer and the logs not faithfully representing data that was sent over the wire, or from some future functionality that doesn't exist yet where this could become a corner case. I can accept it with the documentation and tests as they are and the fact that `SliceSet[T](nil)` and `SliceRemove[T]()` are almost identical for all functionality and most crucially for marshalling, but it still gives me some unease.

If we're really reticent to change the `SliceSet` signature to include an error to exclude `SliceSet[T](nil)` another possibility (albeit a really aggressive one) would be to have a call w/ `nil` create a panic.

Reply:
Essentially, we need to decide whether we want to support the `nil` value for set operations and, if not, the mechanism for disallowing it. For the sake of ergonomics, I'd really like to avoid an error return if possible. And panicking seems a little dangerous. I can think of at least two other options for disallowing `nil` slice-set operations:

1. Return the equivalent of `SliceRemove[T]()` in that case. (This would obviate the need for `SliceRemoveOrSet`.)
2. Allocate an empty array in that case, i.e. `SliceSet[T](nil)` is equivalent to `SliceSet([]T{})`.

This is a tough call. I think I'll make a Nitty Gritty topic on this, unless any of these options jumps out to you as obviously the best.

Reply:
I like option 1. It's less magic than turning a nil slice into an empty one and I think it also would almost always reflect the user's expectation.